pkg/service/v0/service.go

```diff
@@ -2,73 +2,311 @@ package service
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
 
-	"github.com/owncloud/ocis-accounts/pkg/account"
+	"github.com/CiscoM31/godata"
+	"github.com/blevesearch/bleve"
+	"github.com/blevesearch/bleve/search/query"
+	"github.com/gofrs/uuid"
+	"github.com/golang/protobuf/ptypes/empty"
+	mclient "github.com/micro/go-micro/v2/client"
 	"github.com/owncloud/ocis-accounts/pkg/config"
 	"github.com/owncloud/ocis-accounts/pkg/proto/v0"
+	"github.com/owncloud/ocis-accounts/pkg/provider"
 	olog "github.com/owncloud/ocis-pkg/v2/log"
-	mclient "github.com/micro/go-micro/v2/client"
 	settings "github.com/owncloud/ocis-settings/pkg/proto/v0"
+	"github.com/rs/zerolog/log"
+	"github.com/tredoe/osutil/user/crypt"
+
+	// register crypt functions
+	_ "github.com/tredoe/osutil/user/crypt/apr1_crypt"
+	_ "github.com/tredoe/osutil/user/crypt/md5_crypt"
+	_ "github.com/tredoe/osutil/user/crypt/sha256_crypt"
+	_ "github.com/tredoe/osutil/user/crypt/sha512_crypt"
 )
 
 // New returns a new instance of Service
 func New(cfg *config.Config) Service {
-	s := Service{
-		Config: cfg,
+	// read all user and group records
+
+	// for now recreate index on every start
+	os.RemoveAll(filepath.Join(cfg.Server.AccountsDataPath, "index.bleve"))
+	os.MkdirAll(filepath.Join(cfg.Server.AccountsDataPath, "accounts"), 0700)
+
+	mapping := bleve.NewIndexMapping()
+	// TODO don't bother to store fields as we will load the account from disk
+	index, err := bleve.New(filepath.Join(cfg.Server.AccountsDataPath, "index.bleve"), mapping)
+	if err != nil {
+		panic(err)
+	}
+	f, err := os.Open(filepath.Join(cfg.Server.AccountsDataPath, "accounts"))
+	if err != nil {
+		log.Error().Err(err).Str("dir", filepath.Join(cfg.Server.AccountsDataPath, "accounts")).Msg("could not open acconts folder")
+		panic(err)
+	}
+	list, err := f.Readdir(-1)
+	f.Close()
```

Comment:
Shouldn't this be `defer f.Close()`?

Reply:
nah no need for `defer` here. The file descriptor is not used beyond this point nor on a goroutine, so the early it gets freed the better